File: neet_code_all_cpp/array_hash/remove_element.cpp
#include <iostream>
#include <vector>

// 模範解答
class SolutionAns {
public:
    int removeElement(std::vector<int>& nums, int val) {
        int n = (int)nums.size();
        int count=0;
        for(int i = 0;i < n; ++i)
        {
            if(nums[i]!= val)
            {
                std::swap(nums[i], nums[count]);
                count++;
            }
        }
        
        return count;
    }

    // AC
    int removeElement2(std::vector<int>& nums, int val) {
        int k = 0;
        for(int i = 0; i < nums.size(); ++i) {
            if(nums[i] != val) {
                nums[k] = nums[i];
                ++k;
            }
        }

        return k;
    }

    // AC
    int removeElement3(std::vector<int>& nums, int val) {
        int k = 0;
        int i = 0;
        while(i < nums.size()) {
            if(nums[i] == val) {
                ++i;
            } else {
                nums[k] = nums[i];
                ++k;
                ++i;
            }
        }

        return k;
    }
};

void show_resutl(std::vector<int>& result) {
    for(auto&v: result) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
}

int main(void) {
    using Case = std::pair<std::vector<int>, int>;
    Case case_1 = {{3,2,2,3}, 3};
    Case case_2 = {{0,1,2,2,3,0,4,2}, 2};

    SolutionAns s_ans;

    auto res_1 = case_1.first;
    std::cout << s_ans.removeElement(res_1, case_1.second) << std::endl;
    show_resutl(res_1);

    auto res_2 = case_2.first;
    std::cout << s_ans.removeElement(res_2, case_2.second) << std::endl;
    show_resutl(res_2);

    res_1 = case_1.first;
    std::cout << s_ans.removeElement2(res_1, case_1.second) << std::endl;
    show_resutl(res_1);

    res_2 = case_2.first;
    std::cout << s_ans.removeElement2(res_2, case_2.second) << std::endl;
    show_resutl(res_2);

    res_1 = case_1.first;
    std::cout << s_ans.removeElement3(res_1, case_1.second) << std::endl;
    show_resutl(res_1);

    res_2 = case_2.first;
    std::cout << s_ans.removeElement3(res_2, case_2.second) << std::endl;
    show_resutl(res_2);
}

File: neet_code_all/src/array_hash/remove_element.rs
// 解けなかった
struct Solution {}
impl Solution {
    pub fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
        /*
        for i in 0..nums.len() {
            if nums[i] == val {
                for j in i..nums.len() - 1 {
                    nums[j] = nums[j + 1];
                }
                println!("nums: {:?}", nums);
                //nums.pop();
            }
        }
        */

        let mut i = 0;
        let mut j = 0;
        while i < nums.len() {
            if nums[i] == val {
                j = i;
                while nums[j] == val {
                    j += 1;
                }

                for k in i..nums.len() {
                    if k + j - i < nums.len() {
                        nums[k] = nums[k + j - i];
                    } else {
                        break;
                    }
                }
            }

            i += 1;
        }

        nums.len() as i32
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    // 解1
    fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
        while let Some(idx) = nums.iter().position(|v| *v == val) {
            nums.swap_remove(idx);
        }

        nums.len() as i32
    }

    // 解2
    pub fn remove_element_2(nums: &mut Vec<i32>, val: i32) -> i32 {
        let mut k = 0;
        for i in 0..nums.len() {
            if nums[i] != val {
                nums[k] = nums[i];
                k += 1;
            }
        }

        k as i32
    }

    // 解3
    // AC
    pub fn remove_element_3(nums: &mut Vec<i32>, val: i32) -> i32 {
        let mut k = 0;
        let mut i = 0;
        while i < nums.len() {
            if nums[i] == val {
                i += 1;
            } else {
                nums[k] = nums[i];
                k += 1;
                i += 1;
            }
        }

        k as i32
    }
}

// C++の模範解答より
// Rustの解2に近い
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    // AC
    pub fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
        let n = nums.len();
        let mut result = 0;
        for i in 0..n {
            if nums[i] != val {
                nums.swap(i, result);
                result += 1;
            }
        }

        result as i32
    }
}

fn main() {
    let case_1 = (vec![3, 2, 2, 3], 3);
    // => 2, nums = [2, 2, _, _]
    let case_2 = (vec![0, 1, 2, 2, 3, 0, 4, 2], 2);
    // => 5, nums = [0, 1, 4, 0, 3, _, _, _]
    let case_3 = (vec![3, 3, 2, 2, 3], 3);

    /*
    let mut res_1 = case_1.0.clone();
    println!("case_1: {}", Solution::remove_element(&mut res_1, case_1.1));
    println!("{:?}", res_1);

    let mut res_2 = case_2.0.clone();
    println!("case_2: {}", Solution::remove_element(&mut res_2, case_2.1));
    println!("{:?}", res_2);

    let mut res_3 = case_3.0.clone();
    println!("case_3: {}", Solution::remove_element(&mut res_3, case_3.1));
    println!("{:?}", res_3);
    */

    let mut res_1 = case_1.0.clone();
    println!(
        "case_1: {}",
        SolutionAns::remove_element(&mut res_1, case_1.1)
    );
    println!("{:?}", res_1);

    let mut res_2 = case_2.0.clone();
    println!(
        "case_2: {}",
        SolutionAns::remove_element(&mut res_2, case_2.1)
    );
    println!("{:?}", res_2);

    let mut res_3 = case_3.0.clone();
    println!(
        "case_3: {}",
        SolutionAns::remove_element(&mut res_3, case_3.1)
    );
    println!("{:?}", res_3);

    let mut res_1 = case_1.0.clone();
    println!(
        "case_1: {}",
        SolutionAns::remove_element_2(&mut res_1, case_1.1)
    );
    println!("{:?}", res_1);

    let mut res_2 = case_2.0.clone();
    println!(
        "case_2: {}",
        SolutionAns::remove_element_2(&mut res_2, case_2.1)
    );
    println!("{:?}", res_2);

    // C++
    let mut res_1 = case_1.0.clone();
    println!(
        "case_1: {}",
        SolutionAnsCpp::remove_element(&mut res_1, case_1.1)
    );
    println!("{:?}", res_1);

    let mut res_2 = case_2.0.clone();
    println!(
        "case_2: {}",
        SolutionAnsCpp::remove_element(&mut res_2, case_2.1)
    );
    println!("{:?}", res_2);

    let mut res_3 = case_3.0.clone();
    println!(
        "case_3: {}",
        SolutionAnsCpp::remove_element(&mut res_3, case_3.1)
    );
    println!("{:?}", res_3);
}
